kernel/arch/dreamcast/sound/snd_iface.c

```diff
@@ -66,57 +61,53 @@ int snd_init(void) {
 void snd_shutdown(void) {
     if(initted) {
         spu_disable();
-        sem_destroy(&sem_qram);
         snd_mem_shutdown();
         initted = 0;
     }
 }
 
 /* Submit a request to the SH4->AICA queue; size is in uint32's */
 int snd_sh4_to_aica(void *packet, uint32_t size) {
-    uint32_t  qa, bot, start, top, *pkt32, cnt;
-    assert_msg(size < 256, "SH4->AICA packets may not be >256 uint32's long");
+    uint32_t qa, bot, start, top, *pkt32, cnt;
+    assert_msg(size < AICA_CMD_MAX_SIZE, "SH4->AICA packets may not be >256 uint32's long");
 
-    sem_wait(&sem_qram);
+    g2_ctx_t ctx = g2_lock();
 
     /* Set these up for reference */
     qa = SPU_RAM_UNCACHED_BASE + AICA_MEM_CMD_QUEUE;
-    assert_msg(g2_read_32(qa + offsetof(aica_queue_t, valid)), "Queue is not yet valid");
+    assert_msg(g2_read_32_raw(qa + offsetof(aica_queue_t, valid)), "Queue is not yet valid");
 
-    bot = SPU_RAM_UNCACHED_BASE + g2_read_32(qa + offsetof(aica_queue_t, data));
-    top = bot + g2_read_32(qa + offsetof(aica_queue_t, size));
-    start = bot + g2_read_32(qa + offsetof(aica_queue_t, head));
-    pkt32 = (uint32 *)packet;
+    bot = SPU_RAM_UNCACHED_BASE + g2_read_32_raw(qa + offsetof(aica_queue_t, data));
+    top = bot + g2_read_32_raw(qa + offsetof(aica_queue_t, size));
+    start = bot + g2_read_32_raw(qa + offsetof(aica_queue_t, head));
+    pkt32 = (uint32_t *)packet;
     cnt = 0;
 
-    while(size > 0) {
+    while(size-- > 0) {
```

Comment:
Unrelated cosmetic change, if you really want this, do it in a separate commit.

Reply:
Are you sure? GCC can do this loop in another way, by using `dt` for example.
Although he will probably figure it out himself, but still.

Reply:
GCC will absolutely know how to optimize that properly.